src/python/pants/backend/python/macros/poetry_requirements.py

```diff
@@ -167,62 +167,92 @@ def non_pants_project_abs_path(self, path: str) -> Path | None:
         rev_lookup = produce_match("#", attributes.get("rev"))
         branch_lookup = produce_match("@", attributes.get("branch"))
         tag_lookup = produce_match("@", attributes.get("tag"))
 
-        return f"{proj_name} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
-
-    version_lookup = attributes.get("version")
+        base = f"{proj_name}{extras_str} @ git+{git_lookup}{tag_lookup}{branch_lookup}{rev_lookup}"
 
     path_lookup = attributes.get("path")
     if path_lookup is not None:
         non_pants_project_abs_path = pyproject_toml.non_pants_project_abs_path(path_lookup)
         if non_pants_project_abs_path:
-            return f"{proj_name} @ file://{non_pants_project_abs_path}"
-        # An internal path will be handled by normal Pants dependencies and dependency inference;
-        # i.e.: it never represents a third party requirement.
-        return None
+            base = f"{proj_name}{extras_str} @ file://{non_pants_project_abs_path}"
+        else:
+            # An internal path will be handled by normal Pants dependencies and dependency inference;
+            # i.e.: it never represents a third party requirement.
+            return None
 
     url_lookup = attributes.get("url")
     if url_lookup is not None:
-        return f"{proj_name} @ {url_lookup}"
+        base = f"{proj_name}{extras_str} @ {url_lookup}"
 
+    version_lookup = attributes.get("version")
     if version_lookup is not None:
-        markers_lookup = produce_match(";", attributes.get("markers"))
-        fp = str(pyproject_toml.toml_relpath)
-        python_lookup = parse_python_constraint(attributes.get("python"), fp)
-        version_parsed = parse_str_version(proj_name, version_lookup, fp)
-        return (
-            f"{version_parsed}"
-            f"{markers_lookup}"
-            f"{' and ' if python_lookup and markers_lookup else (';' if python_lookup else '')}"
-            f"{python_lookup}"
+        base = parse_str_version(proj_name, version_lookup, fp, extras_str)
+
+    if len(base) == 0:
+        raise AssertionError(
+            f"{proj_name} is not formatted correctly; at minimum provide either a version, url, path "
+            "or git location for your dependency. "
         )
 
-    raise AssertionError(
-        f"{proj_name} is not formatted correctly; at minimum provide either a version, url, path "
-        "or git location for your dependency. "
-    )
+    return add_markers(base, attributes, fp)
 
 
 def parse_single_dependency(
     proj_name: str,
-    attributes: str | dict[str, Any] | list[dict[str, Any]],
+    attributes: str | Mapping[str, Any] | Sequence[Mapping[str, Any]],
     pyproject_toml: PyProjectToml,
 ) -> Iterator[Requirement]:
+
+    # TODO(Liam Wilson): Type hints are a bit of a mess at the moment; I think the best way to
+    # clean this up is a TypedDict for the attributes argument; almost everything else loses a
+    # decent amount of information.
```

Comment:
As discussed, I do think this would be good, mostly from a code being self-documentating persective. That way, future readers don't need to spend as much time as you had to to closely understand the Poetry spec.